File: SampleAppObjC/SampleAppObjC/LocalAuthViewController.m
//
//  LocalAuthViewController.m
//  WhiteLabelDemoApp
//
//  Created by ani on 12/8/17.
//  Copyright © 2017 LaunchKey. All rights reserved.
//

#import "LocalAuthViewController.h"

@interface LocalAuthViewController ()
{
    NSString *localAuthRequestName;
    int expirationDuration;
}

@end

@implementation LocalAuthViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.navigationItem.title = @"Test Local Auth";
    
    _tfCountInput.enabled = NO;
    _switchKnowledge.enabled = NO;
    _switchInherence.enabled = NO;
    _switchPossession.enabled = NO;
    localAuthRequestName = @"";
    expirationDuration = 60;
    
    [_switchCount addTarget:self
                         action:@selector(stateChangedCount:) forControlEvents:UIControlEventValueChanged];
    [_switchType addTarget:self
                         action:@selector(stateChangedType:) forControlEvents:UIControlEventValueChanged];
}

-(void)viewWillAppear:(BOOL)animated
{
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(deviceUnlinked) name:deviceUnlinked object:nil];
}

-(void)deviceUnlinked
{
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.navigationController popViewControllerAnimated:NO];
    });
}

#pragma mark - Switch Methods
- (void)stateChangedCount:(UISwitch *)switchState
{
    if ([switchState isOn])
    {
        _switchType.enabled = NO;
        _tfCountInput.enabled = YES;
    }
    else
    {
        _switchType.enabled = YES;
        _tfCountInput.enabled = NO;
    }
}

- (void)stateChangedType:(UISwitch *)switchState
{
    if ([switchState isOn])
    {
        _switchCount.enabled = NO;
        _switchKnowledge.enabled = YES;
        _switchInherence.enabled = YES;
        _switchPossession.enabled = YES;
    }
    else
    {
        _switchCount.enabled = YES;
        _switchKnowledge.enabled = NO;
        _switchInherence.enabled = NO;
        _switchPossession.enabled = NO;
    }
}

#pragma mark - Generate Local Auth Pressed
- (IBAction)btnGenerateLocalAuthPressed:(id)sender
{
    localAuthRequestName = _tfLARName.text;
    [[LocalAuthManager sharedManager] setTitle:localAuthRequestName];
    if(_tfExpirationDuration.text != nil)
        [[LocalAuthManager sharedManager] setExpiration:[_tfExpirationDuration.text intValue]];
    else
        [[LocalAuthManager sharedManager] setExpiration:60];
        
    if([_switchCount isOn] || [_switchType isOn])
    {
        LKPolicy *policy = [LKPolicy new];
        
        // Check type of Policy to be built
        if([_switchCount isOn])
        {
            policy = [LKPolicy makeWithCountBuilder:^(LKPolicyByCount *builder){
                builder.countTotal = [_tfCountInput.text intValue];
            }];
        }
        else if([_switchType isOn])
        {
            policy = [LKPolicy makeWithTypeBuilder:^(LKPolicyByType *builder)
                      {
                          builder.knowledge = [_switchKnowledge isOn];
                          builder.inherence = [_switchInherence isOn];
                          builder.possession = [_switchPossession isOn];
                      }];
        }
        
        [self presentLocalAuthWithPolicy:policy];
    }
    else
        [self presentLocalAuthWithPolicy:nil];
}

#pragma mark - Presnet Local Auth
-(void)presentLocalAuthWithPolicy:(LKPolicy *)policy
{
    [[LocalAuthManager sharedManager] presentLocalAuth:self.navigationController withPolicy:policy withCompletion:^(BOOL response, NSError *error)
     {
         [self displayResult:response];
         NSLog(@"local auth error: %@", error);
     }];
}

#pragma mark - Display Result
-(void)displayResult:(BOOL)authenticated
{
     if(authenticated)
     {
         UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Approved"
                                                         message:@"User authenticated"
                                                        delegate:self
                                               cancelButtonTitle:@"OK"
                                               otherButtonTitles:nil];

         [alert show];
     }
     else
     {
         UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Denied"
                                                         message:@"User denied"
                                                        delegate:self
                                               cancelButtonTitle:@"OK"
                                               otherButtonTitles:nil];
         
         [alert show];
     }
}

@end

File: SampleAppSwift/SampleAppSwift/LocalAuthViewController.swift
//
//  LocalAuthViewController.swift
//  WhiteLabelDemoAppSwift
//
//  Created by ani on 1/5/18.
//  Copyright © 2018 LaunchKey. All rights reserved.
//

import Foundation

class LocalAuthViewController: UIViewController
{
    
    @IBOutlet weak var tfCountInput: UITextField!
    @IBOutlet weak var switchCount: UISwitch!
    @IBOutlet weak var switchType: UISwitch!
    @IBOutlet weak var switchKnowledge: UISwitch!
    @IBOutlet weak var switchInherence: UISwitch!
    @IBOutlet weak var switchPossession: UISwitch!
    @IBOutlet weak var btnGenerateAuth: UIButton!
    @IBOutlet weak var tfLARName: UITextField!
    @IBOutlet weak var tfExpirationDuration: UITextField!
    
    var localAuthRequestName: String?
    var policy: LKPolicy!
    var expirationDuration : Int32?
    
    override func viewDidLoad()
    {
        
        super.viewDidLoad()
        
        self.title = "Test Local Auth"
        
        tfCountInput.isEnabled = false
        switchKnowledge.isEnabled = false
        switchInherence.isEnabled = false
        switchPossession.isEnabled = false
        localAuthRequestName = ""
        expirationDuration = 60
        
        switchCount.addTarget(self, action: #selector (LocalAuthViewController.stateChangedCount(_:)), for: UIControlEvents.valueChanged)
        
        switchType.addTarget(self, action: #selector (LocalAuthViewController.stateChangedType(_:)), for: UIControlEvents.valueChanged)
    }
    
    override func didReceiveMemoryWarning()
    {
        super.didReceiveMemoryWarning()
    }
    
    override func viewWillAppear(_ animated: Bool)
    {
        NotificationCenter.default.addObserver(self, selector:#selector(deviceIsUnlinked), name: NSNotification.Name(rawValue: deviceUnlinked), object: nil)
    }
    
    func deviceIsUnlinked()
    {
        self.navigationController?.popViewController(animated:false)
    }
    
    func stateChangedCount(_ switchState: UISwitch)
    {
        if switchState.isOn
        {
            switchType.isEnabled = false
            tfCountInput.isEnabled = true
        }
        else
        {
            switchType.isEnabled = true
            tfCountInput.isEnabled = false
        }
    }
    
    func stateChangedType(_ switchState: UISwitch)
    {
        if switchState.isOn
        {
            switchCount.isEnabled = false
            switchKnowledge.isEnabled = true
            switchInherence.isEnabled = true
            switchPossession.isEnabled = true
        }
        else
        {
            switchCount.isEnabled = true
            switchKnowledge.isEnabled = false
            switchInherence.isEnabled = false
            switchPossession.isEnabled = false
        }
    }
    
    @IBAction func btnGenerateLocalAuthPressed(_ sender: Any)
    {
        localAuthRequestName = tfLARName.text
        LocalAuthManager.shared().setTitle(localAuthRequestName)
        let expirationDurationString = self.tfExpirationDuration.text
        if(expirationDurationString != nil)
        {
            LocalAuthManager.shared().setExpiration(Int32(expirationDurationString!)!)
        }
        else
        {
            LocalAuthManager.shared().setExpiration(expirationDuration!)
        }
        
        if switchCount.isOn || switchType.isOn
        {
            // Check type of Policy to be built
            if switchCount.isOn
            {
                policy = LKPolicy.make(countBuilder: { (builder) in
                    let countTotalString = self.tfCountInput.text
                    builder?.countTotal = Int32(countTotalString!)!
                })
            }
            else if switchType.isOn
            {
                policy = LKPolicy.make(typeBuilder: { (builder) in
                    
                    builder?.knowledge = self.switchKnowledge.isOn
                    builder?.inherence = self.switchInherence.isOn
                    builder?.possession = self.switchPossession.isOn
                })
            }
        
            self.presentLocalAuthWithPolicy(policy: policy)
        }
        else
        {
            policy = LKPolicy.make(countBuilder: { (builder) in
                builder?.countTotal = 0
            })
            self.presentLocalAuthWithPolicy(policy: policy)
        }
        
    }
    
    func presentLocalAuthWithPolicy(policy: LKPolicy)
    {
        LocalAuthManager.shared().presentLocalAuth(self.navigationController, with: policy, withCompletion: { (localAuthResponse, error) in
            
            self.displayResult(authenticated: localAuthResponse)
            print("local auth error: \(error)")
        })
    }
    
    func displayResult(authenticated: Bool)
    {
        if(authenticated)
        {
            let alert = UIAlertView()
            alert.title = "Approved"
            alert.message = "User authenticated"
            alert.addButton(withTitle: "OK")
            alert.show()
        }
        else
        {
            let alert = UIAlertView()
            alert.title = "Denied"
            alert.message = "User denied"
            alert.addButton(withTitle: "OK")
            alert.show()
        }
    }
    
}
